users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py

```diff
@@ -0,0 +1,128 @@
+import functools
+import torchaudio
+import torch
+import numpy as np
+import random
+from typing import List, Dict, Optional, Any
+
+class PerturbationFactor:
+    """
+    Class to wrap perturbation factors, e.g. for speed or tempo perturbation.
+    """
+
+    def __init__(self, prob, minimum, maximum):
+        self.prob = prob
+        self.min = minimum
+        self.max = maximum
+
+
+class WaveformPerturbation:
+    """
+    This class enables the perturbation of audio waveforms by applying a variety of transformations such as speed and tempo modification,
+    SoX effects, codec application, and pre-emphasis filtering. 
+
+    :param speed: A dictionary that specifies the parameters for speed perturbation. 
+        - 'prob' (float): The probability of applying speed perturbation. 
+        - 'minimum' (float): The minimum factor by which the audio speed will be decreased.
+        - 'maximum' (float): The maximum factor by which the audio speed will be increased.
+        Example: {"prob": 0.6, "minimum": 0.88, "maximum": 1.12}
+
+    :param tempo: A dictionary specifying the parameters for tempo perturbation.
+        - 'prob' (float): The probability of applying tempo perturbation. 
+        - 'minimum' (float): The minimum factor by which the audio tempo will be decreased.
+        - 'maximum' (float): The maximum factor by which the audio tempo will be increased.
+        Example: {"prob": 0.6, "minimum": 0.83, "maximum": 1.17}
+
+    :param sox_effects: A list of dictionaries, each dictionary representing a SoX effect.
```

Comment:
I think it is a list of lists, isn't it?

Reply:
I don't think so.

Reply:
I do think so :smile:  See the [docs](https://pytorch.org/audio/main/generated/torchaudio.sox_effects.apply_effects_tensor.html).

Reply:
I see.